### ISSUE TITLE
[ERROR] failed to sync service: output= err="signal: killed" (lifecycle sidecar)

### ISSUE THREAD

**Issue by kpurdon** [hashicorp/consul-k8s#295]
Since upgrading via the helm chart v.0.23.1 I've been seeing the following in the lifecycle sidecar pods very often (multiple times per minute): `[ERROR] failed to sync service: output= err="signal: killed"`. This seems to be resulting in eventual container restarts (I've seen ~7 today).

<img width="1413" alt="Screen Shot 2020-07-15 at 8 06 26 PM" src="https://user-images.githubusercontent.com/5130441/87618337-d5c54e00-c6d6-11ea-9b76-a8887552f713.png">

I don't believe this is affecting performance at all but may point to some underlying issue. I alert on container restarts as they often point to performance or other issues in our system. Resource usage seems sane at 60-80% for both memory/cpu and the parent pods/containers are working as expected.

I've attached a chunk of logs, please let me know if I can do any additional debugging or if this is expected behavior.

[connect-lifecycle-sidecar-logs.txt](https://github.com/hashicorp/consul-k8s/files/4928927/connect-lifecycle-sidecar-logs.txt)

**Comment by kpurdon** [comment]
Actually I am seeing memory usage spike to 100%, but that might be a separate issue. Happy to file that separately if it makes sense.

<img width="1399" alt="Screen Shot 2020-07-16 at 9 06 42 AM" src="https://user-images.githubusercontent.com/5130441/87688318-114a3180-c744-11ea-8864-5421d1dfeefc.png">

**Comment by kpurdon** [comment]
And finally, after some more digging I do see: 

```
    Last State:     Terminated
      Reason:       OOMKilled
      Exit Code:    137
```

in the container status logs

**Comment by kpurdon** [comment]
This ended up causing a service disruption for me, so I built a [custom consul-k8s image](https://hub.docker.com/layers/kpurdon/consul-k8s/v0.17.1/images/sha256-401eec35bdcbc775146b5c2a998d5817cd66c6ce3fbc804ec59009724561d793?context=repo) with the following changes: 

```
diff --git a/connect-inject/lifecycle_sidecar.go b/connect-inject/lifecycle_sidecar.go
index cc3f509..352fa0e 100644
--- a/connect-inject/lifecycle_sidecar.go
+++ b/connect-inject/lifecycle_sidecar.go
@@ -10,8 +10,8 @@ import (
 const (
        lifecycleContainerCPULimit      = "20m"
        lifecycleContainerCPURequest    = "20m"
-       lifecycleContainerMemoryLimit   = "25Mi"
-       lifecycleContainerMemoryRequest = "25Mi"
+       lifecycleContainerMemoryLimit   = "50Mi"
+       lifecycleContainerMemoryRequest = "50Mi"
 )

 func (h *Handler) lifecycleSidecar(pod *corev1.Pod) corev1.Container {
```

**Comment by adilyse** [comment]
Hi @kpurdon,

We're planning to make these limits configurable so that folks can adjust them easily if they run into issues. We'll keep you updated on that progress.

Unfortunately, we're still unable to reproduce the issue on our test environments, so the detailed information and graphs are super helpful! To understand your specific case better:

* What version of Kubernetes are you running?
* Are you running on a managed version, bare metal, or on a test environment locally?
* What size are the nodes and how many?
* Are there any resource restrictions in place? Are resource limits required?
* Are there a lot of things deployed that are competing for resources?
* Is there anything else about your setup that might contribute to these issues?

**Comment by kpurdon** [comment]
> What version of Kubernetes are you running?

1.16.9-gke.6

> Are you running on a managed version, bare metal, or on a test environment locally?

GKE

> What size are the nodes and how many?

`n2-highcpu-16` (16 CPU, 16GB mem) and ~10 nodes (auto-scale from 8-12 usually)

> Are there any resource restrictions in place? Are resource limits required?

Not sure exactly what you're looking for here. The various deployments that end up using connect do have requests/limits for resource usage.

> Are there a lot of things deployed that are competing for resources?

There are only 3 deployments using connect and they all have the capacity to grow and the cluster has the ability to scale up as needed w/ resource requests.

> Is there anything else about your setup that might contribute to these issues?

Not that I can think of, pretty standard deployments on GKE.

**Comment by lkysow** [comment]
@kpurdon have you had any issues with it set to 50Mi? From your graph it looks like it just occasionally peaks over 25Mi so maybe 30 or 40Mi might work?

**Comment by kpurdon** [comment]
50Mi has been solid w/ no issues. I intentionally picked a high number (since it's still relatively low) to make sure I was clear. However, I agree that 30-40 Mi would probably end up being fine.

**Comment by Hkyx** [comment]
Got a problem with the same error when connecting ambassador to consul. 
it seems the lifecycle-sidecar show this
```
2020-07-30T19:10:23.742Z [INFO]  successfully synced service: output="Registered service: qotm-sidecar-proxy
Registered service: qotm"
2020-07-30T19:10:39.642Z [INFO]  successfully synced service: output="Registered service: qotm-sidecar-proxy
Registered service: qotm"
2020-07-30T19:10:56.241Z [INFO]  successfully synced service: output="Registered service: qotm-sidecar-proxy
Registered service: qotm"
2020-07-30T19:11:10.745Z [ERROR] failed to sync service: output= err="signal: killed"
```
Tried last version of the helm chart and the version 0.22.0 too, same problem. 
Kubernetes Version: Azure AKS 1.84.4 with Azure Cni Network.

**Comment by lkysow** [comment]
@Hkyx you'll need to use the version of Helm in `master` and this Docker image: `hashicorpdev/consul-k8s:3e25838` until we can get a release out (coming soon).

```yaml
global:
  imageK8S: hashicorpdev/consul-k8s:3e25838
```

**Comment by Hkyx** [comment]
thank @lkysow  ! it works, waiting the release :)

**Comment by adilyse** [comment]
We've released [`consul-k8s v0.18.0`](https://github.com/hashicorp/consul-k8s/blob/master/CHANGELOG.md#0180-july-30-2020) and [`consul-helm v0.24.0`](https://github.com/hashicorp/consul-helm/blob/master/CHANGELOG.md#0240-july-31-2020) that make these settings configurable to address this issue, as well as bumping the lifecycle sidecar memory limit default to `50Mi`.

We're still investigating whether there's a larger underlying issue, but this should unblock folks while we figure that out.

**Comment by kpurdon** [comment]
Upgraded today and as expected things are looking fine. Up to you all if you want to leave this issue open or close it. Having configurable limits solves my issue and resource usage does seem stable for me.

**Comment by lkysow** [comment]
Since resource limits are now configurable and the current settings seem to be working we're going to close this.
